handlers/authentication.go

```diff
@@ -84,6 +86,41 @@ func getTokenStringFromRequest(r *http.Request) string {
 	return tokenString
 }
 
+func getTokenStringFromHeader(r *http.Request) *api.AuthInfo {
+	tokenString := "" // Default to no token.
+
+	// Extract token from the Authorization HTTP header sent from the reverse proxy
+	if headerValue := r.Header.Get("Authorization"); strings.Contains(headerValue, "Bearer") {
+		tokenString = strings.TrimPrefix(headerValue, "Bearer ")
+	}
+
+	authInfo := &api.AuthInfo{Token: tokenString}
+
+	impersonationHeader := r.Header.Get("Impersonate-User")
+	if len(impersonationHeader) > 0 {
+		//there's an impersonation header, lets make sure to add it
+		authInfo.Impersonate = impersonationHeader
+
+		//Check for impersonated groups
+		if groupsImpersonationHeader := r.Header["Impersonate-Group"]; len(groupsImpersonationHeader) > 0 {
+			authInfo.ImpersonateGroups = groupsImpersonationHeader
+		}
+
+		//check for extra fields
+		for headerName, headerValues := range r.Header {
+			if strings.HasPrefix(headerName, "Impersonate-Extra-") {
+				extraName := headerName[len("Impersonate-Extra-"):]
+				if authInfo.ImpersonateUserExtra == nil {
+					authInfo.ImpersonateUserExtra = make(map[string][]string)
```

Comment:
This is just a comment: Just making sure users elsewhere of authInfo will know to check for both nil and non-nil values of ImpersonateUserExtra.  I haven't looked closely, but just wanted to point out this value may be nil so code that reads this value may need to check for nil rather than, say, a length of 0.

Reply:
i checked, this particular code was just for lazy initialization, so to say "if we haven't initialed, create the map"